File: MissingNumberSortedArray.py
# T(n)=O(log n)
# S(n) = O(1)

def search(arr):
    low=0
    high=len(arr)-1
   
    while(low<=high):
        mid=low+(high-low)//2
        if arr[mid]==mid+1:
            low=mid+1
            mid=mid+1
        else:
            high=mid-1
    print(mid+1)
search([1, 2, 3, 4, 5, 6, 8, 9])

File: MissingNumber_SortedArray.java
//MissingNumber

//TimeComplexity O(n)
//SpaceComplexity O(1)

class Solution {
    public int missingNumber(int[] nums) {
        
        
        if(nums[0] != 0)return 0;
       
        if(nums[nums.length-1]- (nums.length-1) == 0)return nums.length;
        for(int i = 0 ; i < nums.length; i++){
            if(nums[i]-i != 0){
                return i;
            } 
        }
        return -1;
    }
}

//TimeComplexity logn
//SpaceComplexity O(1)

class Solution {
    public int missingNumber(int[] nums) {
        Arrays.sort(nums);
        
        if(nums[0] != 0)return 0;
       
        if(nums[nums.length-1]- (nums.length-1) == 0)return nums.length;
        
        int low = 0;
        int high = nums.length-1;
        
        while(low < high){
            int mid = low + (high - low)/2;
            
            if(nums[mid]-mid == 1){
                high = mid;
            }else{
                low = mid+1;
            }
        }
        return high;
    }
}
